modules/tinymce/src/core/main/ts/html/ExtractInvalidNodes.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Fun } from '@ephox/katamari';
+
+import AstNode from '../api/html/Node';
+import Schema from '../api/html/Schema';
+import Tools from '../api/util/Tools';
+import { hasOnlyChild, isEmpty } from './ParserUtils';
+
+const removeOrUnwrapInvalidNode = (node: AstNode, schema: Schema, originalNodeParent: AstNode = node.parent) => {
+  if (schema.getSpecialElements()[node.name]) {
+    node.empty().remove();
+  } else {
+    // are the children of `node` valid children of the top level parent?
+    // if not, remove or unwrap them too
+    const children = node.children();
+    for (const childNode of children) {
+      if (!schema.isValidChild(originalNodeParent.name, childNode.name)) {
+        removeOrUnwrapInvalidNode(childNode, schema, originalNodeParent);
+      }
+    }
+    node.unwrap();
+  }
+};
+
+const extractInvalidNodes = (nodes: AstNode[], schema: Schema, onCreate: (newNode: AstNode) => void = Fun.noop): void => {
```

Comment:
Should this be called `extract` as that keeps confusing me here since it doesn't return anything? Maybe call it `cleanInvalidNodes` instead?

Reply:
Also maybe just call the module `InvalidNodes` as that's what it's purpose is even if it only exposes the one function atm.